domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt

```diff
@@ -161,101 +192,310 @@ class TopicListController @Inject constructor(
       .build()
   }
 
-  private fun createOngoingStoryListFromProgress(
+  private fun createUpcomingTopicSummaryFromJson(
+    topicId: String,
+    jsonObject: JSONObject
+  ): UpcomingTopic {
+    val upcomingTopic = UpcomingTopic.newBuilder()
+    var totalChapterCount = 0
+    val storyData = jsonObject.getJSONArray("canonical_story_dicts")
+    for (i in 0 until storyData.length()) {
+      totalChapterCount += storyData
+        .getJSONObject(i)
+        .getJSONArray("node_titles")
+        .length()
+    }
+    val topicPlayAvailability = if (jsonObject.getBoolean("published")) {
+      TopicPlayAvailability.newBuilder().setAvailableToPlayNow(true).build()
+    } else {
+      TopicPlayAvailability.newBuilder().setAvailableToPlayInFuture(true).build()
+    }
+
+    return upcomingTopic.setTopicId(topicId)
+      .setName(jsonObject.getString("topic_name"))
+      .setVersion(jsonObject.optInt("version"))
+      .setTopicPlayAvailability(topicPlayAvailability)
+      .setLessonThumbnail(createTopicThumbnail(jsonObject))
+      .build()
+  }
+
+  private fun computePromotedActivityList(
+    topicProgressList: List<TopicProgress>
+  ): PromotedActivityList {
+    val promotedActivityListBuilder = PromotedActivityList.newBuilder()
+    if (topicProgressList.isNotEmpty()) {
+      promotedActivityListBuilder.promotedStoryList = computePromotedStoryList(topicProgressList)
+      if (promotedActivityListBuilder.promotedStoryList.getTotalPromotedStoryCount() == 0) {
+        promotedActivityListBuilder.comingSoonTopicList = computeComingSoonTopicList()
+      }
+    }
+    return promotedActivityListBuilder.build()
+  }
+
+  private fun computePromotedStoryList(topicProgressList: List<TopicProgress>): PromotedStoryList {
+    val promotedStoryListBuilder = PromotedStoryList.newBuilder()
+    promotedStoryListBuilder
+      .addAllUpTo(
+        populateRecentlyPlayedStories(topicProgressList),
+        PromotedStoryList.Builder::addAllRecentlyPlayedStory,
+        limit = 3
```

Comment:
@BenHenning if we have tablet device we are showing 4 items. Please confirm on this, if I need to set limit over here.

Reply:
We should make this independent of the UI (since it's the domain layer). If this is too difficult to reasonably know at this stage, we could alternatively simplify the domain layer a bit by just passing all recently played/older/suggested stories and letting the UI do this limit 3/4 merging logic. That actually might simplify both this code & the corresponding merging code in the UI since then we just need to merge into a single list rather than keeping the three different versions. WDYT?

@rt4914 as well in case you have any thoughts.

Reply:
I do think that this should be manager in `app` layer so that things remain simple and in future we can scale it easily too if needed.

Reply:
@rt4914 and @BenHenning so what is the final thing. Do I need to keep limit in domain layer or I shall remove and manage in UI.

Reply:
`Pass all recently played/older/suggested stories and letting the UI do this limit 3/4 merging logic` - @veena14cs

Reply:
It seems like this has been addressed in the latest changes.